### PR TITLE
[RzIL] Add lifter for Armv8 AArch64

### PR DESCRIPTION
**Pull request by thestr4ng3r**
# SQUASH ME

### CODE REVIEW

File: librz/analysis/arch/arm/arm_il32.c [hunk 1]
```diff
@@ -342,41 +344,6 @@ static RzILOpEffect *update_flags_from_cpsr(RzILOpBitVector *val, bool f, bool s
 	return setf && sets ? SEQ2(sets, setf) : (setf ? setf : sets);
 }
 
-/**
```
Comment (XVilka) [inline librz/analysis/arch/arm/arm_il32.c]: Could you please extract that small refactoring change into a separate PR that we can merge before?

File: librz/analysis/arch/arm/arm_il64.c [hunk 2]
```diff
@@ -0,0 +1,2682 @@
...
+		return NULL;
+	}
+	if (is_wreg(reg)) {
+		v = UNSIGNED(64, v);
+	}
+	return SETG(var, v);
+}
+
+static RzILOpBitVector *arg_mem(RzILOpBitVector *base_plus_disp, cs_arm64_op *op) {
+	if (op->mem.index != ARM64_REG_INVALID) {
```
Comment (XVilka) [inline librz/analysis/arch/arm/arm_il64.c]: Maybe invert the condition here?

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/analysis/arch/arm/arm_il32.c (hunk 1 above)
Comment (thestr4ng3r) [reply librz/analysis/arch/arm/arm_il32.c]: https://github.com/rizinorg/rizin/pull/2394

File: librz/analysis/arch/arm/arm_il64.c [hunk 3]
```diff
@@ -0,0 +1,2682 @@
...
+/**
+ * Capstone: ARM64_INS_MOV, ARM64_INS_MOVZ
+ * ARM: mov, movz
+ */
+static RzILOpEffect *mov(cs_insn *insn) {
+	if (!ISREG(0)) {
+		return NULL;
+	}
+	if (ISIMM(1) && IMM(1) == 0 && !strcmp(insn->mnemonic, "movn")) {
+		// Capstone bug making 0000a012 indistinguishable from 0000a052
```
Comment (XVilka) [inline librz/analysis/arch/arm/arm_il64.c:1614]: Should we open an issue to track this? And add a link here in the comment?

**Review by XVilka — COMMENTED** [review]
Looks good from the code point of view. Amazing job, as always! 💯

File: librz/analysis/arch/arm/arm_il64.c (hunk 3 above)
Comment (thestr4ng3r) [reply librz/analysis/arch/arm/arm_il64.c:1614]: Yes, maybe. I have not checked how it looks after llvm sync though.

Comment (thestr4ng3r) [reply librz/analysis/arch/arm/arm_il64.c:1614]: https://github.com/capstone-engine/capstone/issues/1857

File: librz/analysis/arch/arm/arm_il64.c [hunk 4]
```diff
@@ -0,0 +1,2713 @@
...
+}
+
+static ut32 reg_bits(arm64_reg reg) {
+	if (is_xreg(reg) || reg == ARM64_REG_XZR) {
+		return 64;
+	}
+	if (is_wreg(reg) || reg == ARM64_REG_WZR) {
+		return 32;
+	}
+	return 0;
```
Comment (wargio) [inline librz/analysis/arch/arm/arm_il64.c:232]: maybe add `rz_warn_if_reached()` ?

File: librz/analysis/arch/arm/arm_il64.c [hunk 5]
```diff
@@ -0,0 +1,2713 @@
...
+			addr = ADD(addr, U64(disp));
+		} else if (disp < 0) {
+			addr = SUB(addr, U64(-disp));
+		}
+		return arg_mem(addr, &insn->detail->arm64.operands[n]);
+	}
+	default:
+		break;
+	}
+	return NULL;
```
Comment (wargio) [inline librz/analysis/arch/arm/arm_il64.c:405]: warn_if_reached()

Comment (thestr4ng3r) [reply librz/analysis/arch/arm/arm_il64.c:405]: No, `NULL` is valid here for anything we don't handle yet. It is checked by callers.

File: librz/analysis/arch/arm/arm_il64.c [hunk 6]
```diff
@@ -0,0 +1,2713 @@
...
+	if (!ISREG(0)) {
+		return NULL;
+	}
+#if 0
+	if ((insn->id == ARM64_INS_ADD || insn->id == ARM64_INS_SUB) &&
+		!insn->detail->arm.update_flags && OPCOUNT() == 3 && REGID(1) == ARM64_REG_PC && ISIMM(2)) {
+		// alias for adr
+		return adr(insn, is_thumb);
+	}
+#endif
```
Comment (wargio) [inline librz/analysis/arch/arm/arm_il64.c]: why this is removed?

Comment (thestr4ng3r) [reply librz/analysis/arch/arm/arm_il64.c]: I don't remember, and since there is no comment with more info, no use in keeping in my opinion.

File: librz/analysis/arch/arm/arm_il64.c [hunk 7]
```diff
@@ -0,0 +1,2707 @@
...
+	if (!ISREG(0)) {
+		return NULL;
+	}
+	cs_arm64_op *op = &insn->detail->arm64.operands[1];
+#if CS_API_MAJOR > 4
+	if (op->type != ARM64_OP_SYS || op->sys != ARM64_SYSREG_NZCV) {
+		return NULL;
+	}
+#else
+	if (op->type != ARM64_OP_REG_MRS || op->reg != 0xda10) {
```
Comment (wargio) [inline librz/analysis/arch/arm/arm_il64.c:1763]: wasn't simpler to define `ARM64_SYSREG_NZCV` this on top of the file if `CS_API_MAJOR < 5` ?

File: librz/analysis/p/analysis_arm_cs.c [hunk 8]
```diff
@@ -1755,7 +1755,7 @@ static int analysis_op(RzAnalysis *a, RzAnalysisOp *op, ut64 addr, const ut8 *bu
 			return -1;
 		}
 	}
-	int haa = hackyArmAnal(a, op, buf, len);
+	int haa = hackyArmAnal(a, op, buf, len); // TODO: disable this for capstone 5 after testing that everything works
```
Comment (wargio) [inline librz/analysis/p/analysis_arm_cs.c:1758]: Unrelated to the PR, but this function name shall be renamed.

**Review by wargio — APPROVED** [review]

File: librz/analysis/arch/arm/arm_il64.c (hunk 7 above)
Comment (thestr4ng3r) [reply librz/analysis/arch/arm/arm_il64.c:1763]: That would suggest that it's a stable constant, while it is rather just observation-based, which is the best we can do for cs4.

**Review by XVilka — APPROVED** [review]
